src/mito2/src/worker/handle_alter.rs

```diff
@@ -0,0 +1,137 @@
+use crate::memtable::MemtableBuilderRef;
+use crate::region::version::Version;
+use crate::region::MitoRegionRef;
+use crate::request::{DdlRequest, SenderDdlRequest};
+use crate::worker::{send_result, RegionWorkerLoop};
+
+impl<S> RegionWorkerLoop<S> {
+    pub(crate) async fn handle_alter_request(
+        &mut self,
+        region_id: RegionId,
+        request: RegionAlterRequest,
+        sender: Option<oneshot::Sender<Result<Output>>>,
+    ) {
+        let Some(region) = self.regions.get_region(region_id) else {
+            send_result(sender, RegionNotFoundSnafu { region_id }.fail());
+            return;
+        };
+
+        info!("Try to alter region: {}, request: {:?}", region_id, request);
+
+        let version = region.version();
+        // Checks whether we can alter the region directly.
+        if !can_alter_directly(&version) {
+            // We need to flush all memtables first.
+            info!("Flush region: {} before alteration", region_id);
+
+            // Try to submit a flush task.
+            let task = self.new_flush_task(&region, FlushReason::Alter);
+            if let Err(e) = self.flush_scheduler.schedule_flush(&region, task) {
+                // Unable to flush the region, send error to waiter.
+                send_result(sender, Err(e));
+                return;
+            }
+
+            // Safety: We have requested flush.
+            self.flush_scheduler
+                .add_ddl_request_to_pending(SenderDdlRequest {
+                    region_id,
+                    sender,
+                    request: DdlRequest::Alter(request),
+                });
+
+            return;
+        }
+
+        // Now we can alter the region directly.
+        if let Err(e) =
+            alter_region_schema(&region, &version, request, &self.memtable_builder).await
+        {
+            error!(e; "Failed to alter region schema, region_id: {}", region_id);
+            send_result(sender, Err(e));
+            return;
+        }
+
+        info!("Schema of region {} is altered", region_id);
+
+        // Notifies waiters.
+        send_result(sender, Ok(Output::AffectedRows(0)));
+    }
+}
+
+/// Alter the schema of the region.
+async fn alter_region_schema(
+    region: &MitoRegionRef,
+    version: &Version,
+    request: RegionAlterRequest,
+    builder: &MemtableBuilderRef,
+) -> Result<()> {
+    let new_meta = metadata_after_alteration(&version.metadata, request)?;
+    // Persist the metadata to region's manifest.
+    let change = RegionChange {
+        metadata: new_meta.clone(),
+    };
+    let action_list = RegionMetaActionList::with_action(RegionMetaAction::Change(change));
+    region.manifest_manager.update(action_list).await?;
+
+    // Apply the metadata to region's version.
+    region.version_control.alter_schema(new_meta, builder);
+    Ok(())
+}
+
+/// Checks whether all memtables are empty.
+fn can_alter_directly(version: &Version) -> bool {
```

Comment:
Seems better to move this method to `Version`? Or add a `is_empty()` to `Version`